# Juliaによる数値計算

## 基本的な数学関数

In [1]:
# e^2
exp(2) |> println

# ln(e)
## 自然対数e: exp(1) or ℯ (\euler + <tab>)
log(ℯ) |> println

# sin(π): 数学的には 0 になるが、浮動小数点の誤差により 0 に近い値が返る
## 円周率π: \pi + <tab>
sin(π) |> println

# √3
sqrt(3) |> println

7.38905609893065
1
1.2246467991473532e-16
1.7320508075688772


## Julia の配列・行列

Julia の配列はデフォルトで縦ベクトルとして定義されている

配列はリストのように要素を並べたものだが、すべての要素の型は同じでなければならない

また、一般的なプログラミング言語と異なり、Julia では配列の先頭要素は添字 1 でアクセスする（一般的なプログラミング言語では、配列要素は 0 から始まる）

In [2]:
# 配列定義
a = [2, 3, 5, 7, 8]
display(a)

# 先頭要素: a[1] = 2
println(a[1])

# インデックス 2～3 の要素: [3, 5]
println(a[2:3])

# 最後から一つ前の要素: a[5-1] = a[4] = 7
println(a[end-1])

5-element Vector{Int64}:
 2
 3
 5
 7
 8

2
[3, 5]
7


In [3]:
# UnitRange による数列
## 1～5の等差数列の配列を作成
b = Vector(1:5)

5-element Vector{Int64}:
 1
 2
 3
 4
 5

In [4]:
# UnitRange: start:step:end
## 1.0～2.0, 等差0.2 の数列の配列を作成
c = Vector(1.0:0.2:2.0)

6-element Vector{Float64}:
 1.0
 1.2
 1.4
 1.6
 1.8
 2.0

In [5]:
# 型の確認
typeof(a)

Vector{Int64} (alias for Array{Int64, 1})

In [6]:
# 配列作成時に要素の型を指定
d = Vector{Float64}([1, 2, 3])

3-element Vector{Float64}:
 1.0
 2.0
 3.0

その他、Julia の配列・行列関連操作については [01_tutorial/07_Julia_vector.ipynb](../01_tutorial/07_Julia_vector.ipynb) を参照

### 配列の基本計算
Julia では、配列・行列に対して集計計算を行うための関数が豊富に用意されている

In [7]:
a = Vector(1.0:5.0)
display(a)

# 合計値算出: 1.0 + 2.0 + ... + 5.0
sum(a) |> println

# 最大値算出: 0.5
maximum(a) |> println

# 最小値算出: 1.0
minimum(a) |> println

5-element Vector{Float64}:
 1.0
 2.0
 3.0
 4.0
 5.0

15.0
5.0
1.0


In [8]:
# Statistics 標準パッケージを使えば、平均値や標準偏差などの記述統計用関数を使うことが出来る
using Statistics

# 平均
mean(a) |> println

# 標準偏差: デフォルトでは不偏標準偏差を計算する
std(a) |> println

3.0
1.5811388300841898


In [9]:
# 2次元行列の場合
b = [
    0. 1. 2.;
    3. 4. 5.;
    6. 7. 8.
]
display(b)

# 合計値算出: 0.0 + 1.0 + ... + 8.0
sum(b) |> println

# 最大値算出: 8.0
maximum(b) |> println

# 最小値算出: 0.0
minimum(b) |> println

# 平均: 4.0
mean(b) |> println

3×3 Matrix{Float64}:
 0.0  1.0  2.0
 3.0  4.0  5.0
 6.0  7.0  8.0

36.0
8.0
0.0
4.0


In [10]:
"""
行列演算の場合、次元を指定して演算の方向を決めることが出来る
"""

# 列方向で合計値算出: [(0.0 + 3.0 + 6.0) (1.0 + 4.0 + 7.0) (2.0 + 5.0 + 8.0)]
sum(b; dims=1) |> display

# 行方向で合計値算出: [(0.0 + 1.0 + 2.0); (3.0 + 4.0 + 5.0); (6.0 + 7.0 + 8.0)]
sum(b; dims=2) |> display

1×3 Matrix{Float64}:
 9.0  12.0  15.0

3×1 Matrix{Float64}:
  3.0
 12.0
 21.0

In [11]:
"""
Julia で一要素に対して適用される関数を、配列・行列に対して適用する場合は '.'(ドット演算子)を用いる
"""

a = Vector{Float64}(3:7)
display(a)

# 配列 a の全要素に exp 適用
exp.(a) |> display

# 配列 a の全要素に 3 を乗算
a .* 3 |> display

# 配列 a の全要素に対して > 4 の条件式を適用
(a .> 4) |> display

5-element Vector{Float64}:
 3.0
 4.0
 5.0
 6.0
 7.0

5-element Vector{Float64}:
   20.085536923187668
   54.598150033144236
  148.4131591025766
  403.4287934927351
 1096.6331584284585

5-element Vector{Float64}:
  9.0
 12.0
 15.0
 18.0
 21.0

5-element BitVector:
 0
 0
 1
 1
 1

In [12]:
"""
配列・行列のフィルタリング
"""
# 条件式を各要素に適用した結果の BitVector を利用することで配列のフィルタリングを行うことができる
## a_i > 4 である要素をフィルタリング
a[a .> 4] |> display

# filter 関数を用いることで複雑なフィルタリングが可能
## a_i が偶数である要素をフィルタリング
filter(a_i -> a_i % 2. === 0., a) |> display

3-element Vector{Float64}:
 5.0
 6.0
 7.0

2-element Vector{Float64}:
 4.0
 6.0

In [13]:
# findall 関数を使うと、条件を満たす要素の index を取得できる

# a_i > 4 である要素の index を取得
findall(a .> 4) |> display

# a_i が偶数である要素の index を取得
findall(a_i -> a_i % 2. === 0., a) |> display

# a_i > 4 である要素の index を取得し、その index の要素を a から抽出
## a[a .> 4] と同値
a[findall(a .> 4)] |> display

3-element Vector{Int64}:
 3
 4
 5

2-element Vector{Int64}:
 2
 4

3-element Vector{Float64}:
 5.0
 6.0
 7.0

In [14]:
"""
配列同士の演算
"""
# 内積関数 dot を使うには LinearAlgebra 標準パッケージを呼び出す
using LinearAlgebra

u = collect(0:3)
v = collect(3:6)
display(u)
display(v)

# 配列同士の加算
u + v |> display

# 配列同士の減算
u - v |> display

# 配列同士の乗算: 定義されていないため、各要素同士の乗算とする
u .* v |> display

# 内積
dot(u, v) |> display

# 配列同士の乗算を行い、合計値を算出
## これがすなわち内積となる
u .* v |> sum |> display

4-element Vector{Int64}:
 0
 1
 2
 3

4-element Vector{Int64}:
 3
 4
 5
 6

4-element Vector{Int64}:
 3
 5
 7
 9

4-element Vector{Int64}:
 -3
 -3
 -3
 -3

4-element Vector{Int64}:
  0
  4
 10
 18

32

32

## 疎行列

機械学習においては、ほとんどの要素が 0 であるような行列を扱うことがよくある

このような行列を **疎行列** と呼ぶ

疎行列は通常の2次元配列を用いて計算を行うと（0 の演算ばかりになるため）計算効率もメモリ効率も悪いため、Julia では疎行列専用の型 `SparseMatrixCSC` が用意されている

なお、疎行列に対して通常の行列のことを密行列と呼ぶこともある

疎行列と密行列は、抽象的な数学のレベルでは同じものであるが、実際にコンピュータ上で計算を行う際には、状況に応じて疎行列専用のデータ型を用いると便利なことがある

In [15]:
# 疎行列のデータ型や関連操作は SparseArrays 標準パッケージにまとめられている
## 公式リファレンス: https://docs.julialang.org/en/v1/stdlib/SparseArrays/
using SparseArrays

# 全要素が 0 の 4×5-疎行列を作成
a = spzeros(4, 5)

4×5 SparseMatrixCSC{Float64, Int64} with 0 stored entries:
  ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅ 

In [16]:
# 疎行列の各要素を更新
a[1, 2] = 1
a[1, 4] = 2
a[3, 3] = 3
a[4, 5] = 4
display(a)

4×5 SparseMatrixCSC{Float64, Int64} with 4 stored entries:
  ⋅   1.0   ⋅   2.0   ⋅ 
  ⋅    ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅   3.0   ⋅    ⋅ 
  ⋅    ⋅    ⋅    ⋅   4.0

In [17]:
# 疎行列を蜜行列（Matrix型）に変換
Matrix(a)

4×5 Matrix{Float64}:
 0.0  1.0  0.0  2.0  0.0
 0.0  0.0  0.0  0.0  0.0
 0.0  0.0  3.0  0.0  0.0
 0.0  0.0  0.0  0.0  4.0

In [18]:
# 密行列（Matrix型）から疎行列型のデータを生成
b = sparse([0 0 1. 0; 0 0 2. 0; 0 0 0 3.; 0 0 0 4.; 0 0 0 0])

5×4 SparseMatrixCSC{Float64, Int64} with 4 stored entries:
  ⋅    ⋅   1.0   ⋅ 
  ⋅    ⋅   2.0   ⋅ 
  ⋅    ⋅    ⋅   3.0
  ⋅    ⋅    ⋅   4.0
  ⋅    ⋅    ⋅    ⋅ 

In [19]:
# 疎行列同士の内積
## Julia において数学的な意味での行列の内積は * 演算子で計算可能
### 数学的な意味での行列の内積: (m×n)行列 × (n×k)行列 の内積
c = a * b

4×4 SparseMatrixCSC{Float64, Int64} with 3 stored entries:
  ⋅    ⋅   2.0  8.0
  ⋅    ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅   9.0
  ⋅    ⋅    ⋅    ⋅ 

## Julia による線形代数

Julia における転置行列は `transpose` 関数、もしくは `'` (`LinearAlgebra.Adjoint`) 演算子により求めることができる

なお `'` は、正確にはエルミート行列を求める演算子であり、複素行列の場合は共役転置行列になるため注意が必要である

In [20]:
a = [
    3 1 1;
    1 2 1;
    0 -1 1
]
transpose(a) |> display
a' |> display

3×3 transpose(::Matrix{Int64}) with eltype Int64:
 3  1   0
 1  2  -1
 1  1   1

3×3 adjoint(::Matrix{Int64}) with eltype Int64:
 3  1   0
 1  2  -1
 1  1   1

また、逆行列を求めたい場合は `inv` 関数を用いる

なお、`inv` 関数は内部的に [01-02_matrix.ipynb](./01-02_matrix.ipynb) で紹介したガウスの消去法を用いて逆行列を求めているため、計算効率としては実はあまり良くない

In [21]:
inv(a)

3×3 Matrix{Float64}:
  0.428571  -0.285714  -0.142857
 -0.142857   0.428571  -0.285714
 -0.142857   0.428571   0.714286

ここで、次の連立方程式を数値的に求めることを考える

$$
\begin{align}
3x + y + z &= 1 \\
x + 2y + z &= 2 \\
-y + z &= 3
\end{align}
$$

この方程式を行列式で表すと、

$$
\begin{pmatrix}
3 & 1 & 1 \\
1 & 2 & 1 \\
0 & -1 & 1
\end{pmatrix}
\begin{pmatrix}
x \\ y \\ z
\end{pmatrix}
=
\begin{pmatrix}
1 \\ 2 \\ 3
\end{pmatrix}
$$

となり、$\boldsymbol A = \begin{pmatrix}3 & 1 & 1 \\ 1 & 2 & 1 \\ 0 & -1 & 1\end{pmatrix}$ は正則行列であるため、この逆行列 $\boldsymbol A^{-1}$ を用いて、以下のように解くことが出来る

$$
\begin{align}
\begin{pmatrix}
3 & 1 & 1 \\
1 & 2 & 1 \\
0 & -1 & 1
\end{pmatrix}^{-1}
\begin{pmatrix}
3 & 1 & 1 \\
1 & 2 & 1 \\
0 & -1 & 1
\end{pmatrix}
\begin{pmatrix}
x \\ y \\ z
\end{pmatrix}
&=
\begin{pmatrix}
3 & 1 & 1 \\
1 & 2 & 1 \\
0 & -1 & 1
\end{pmatrix}^{-1}
\begin{pmatrix}
1 \\ 2 \\ 3
\end{pmatrix} \\
\begin{pmatrix}
x \\ y \\ z
\end{pmatrix}
&=
\begin{pmatrix}
  0.428571 & -0.285714 & -0.142857 \\
 -0.142857 &  0.428571 & -0.285714 \\
 -0.142857 &  0.428571 &  0.714286
\end{pmatrix}
\begin{pmatrix}
1 \\ 2 \\ 3
\end{pmatrix} \\
&\approx
\begin{pmatrix}
-0.571429 \\ -0.142857 \\ 2.857143
\end{pmatrix}
\end{align}
$$

In [22]:
inv([3. 1. 1.; 1. 2. 1.; 0. -1. 1.]) * [1.; 2.; 3.]

3-element Vector{Float64}:
 -0.5714285714285714
 -0.14285714285714302
  2.8571428571428568

前述の通り、逆行列を求める計算は効率が良いとは言えない

そのため、連立方程式の解を数値的に求める場合は、**LU分解** というアルゴリズムを使うことが多い

LU分解とは与えられた$n$次正方行列を、置換行列 $\boldsymbol P$、対角成分が 1 の下三角行列 $\boldsymbol L$、上三角行列 $\boldsymbol U$ を使って

$$
\boldsymbol A = \boldsymbol P \boldsymbol L \boldsymbol U
$$

と表すことである

- 置換行列:
    - 各行に 1 である成分がちょうど1つだけあり、ほかは全部 0 であるような行列
- 下三角行列:
    - 対角成分より右上がすべて 0 である行列
- 上三角行列:
    - 対角成分より左下がすべて 0 である行列

$\boldsymbol L$ は対角成分が全て 1 であるという条件もあったため、$\boldsymbol L$ と $\boldsymbol U$ は次のように表される

$$
\boldsymbol L = \begin{pmatrix}1&&&& \\ *&1&&& \\ *&*&1&& \\ \vdots&\vdots&\vdots&\ddots& \\ *&*&*&\cdots&1\end{pmatrix},
\boldsymbol U = \begin{pmatrix}*&*&*&*&* \\ &*&*&*&* \\ &&\ddots&\vdots&\vdots \\ &&&*&* \\ &&&&*\end{pmatrix}
$$

ただし、ここで $*$ は任意の値を表す

このような $\boldsymbol P$, $\boldsymbol L$, $\boldsymbol U$ を係数に持つ連立方程式は高速に解けることが分かっている

もともと解きたかったのは、方程式 $\boldsymbol A \boldsymbol x = \boldsymbol b$ だったため、

$$
\boldsymbol P \boldsymbol L \boldsymbol U \boldsymbol x = \boldsymbol b
$$

を満たす $\boldsymbol x$ を求めれば良いことになる

これを求めるには、次の方程式の解を逐次求めていけば良い

$$
\begin{align}
\boldsymbol P \boldsymbol z &= \boldsymbol b \\
\boldsymbol L \boldsymbol y &= \boldsymbol z \\
\boldsymbol U \boldsymbol x &= \boldsymbol y
\end{align}
$$

上記のようにして求められた $\boldsymbol x$ は、元の方程式の解となる

このそれぞれの方程式は効率的に解くことができ、その計算量は、$n$次正方行列とベクトルの積を計算する計算量とほぼ同じとなる

また、LU分解のための（$\boldsymbol P$, $\boldsymbol L$, $\boldsymbol U$ を求めるための）計算量は $\boldsymbol A^{-1}$ を求める計算量より少ないため、最終的に逆行列を使って計算するよりも効率が良いということになる

In [23]:
# LU分解関数
function lu_parse(xs::Matrix{Float64})
    n = size(xs, 1)
    zs = copy(xs)
    for i = 1 : n
        for j = i + 1 : n
            temp = zs[j, i] / zs[i, i]
            for k = i + 1 : n
                zs[j, k] -= temp * zs[i, k]
            end
            zs[j, i] = temp
        end
    end
    return zs
end

# LU分解された結果から方程式を解く関数
function lu_solve(xs::Matrix{Float64}, ys::Vector{Float64})
    n = size(xs, 1)
    zs = copy(ys)
    # 前進代入
    for i = 2 : n, j = 1 : i - 1
        zs[i] -= xs[i, j] * zs[j]
    end
    # 後退代入
    for i = n : -1 : 1
       for j = i + 1 : n
           zs[i] -= xs[i, j] * zs[j]
       end
       zs[i] /= xs[i, i]
    end
    return zs
end

# 方程式係数行列
A = [3. 1. 1.; 1. 2. 1.; 0. -1. 1.]

# 方程式定数ベクトル（右辺値）
b = [1.; 2.; 3.]

# LU分解
xs = lu_parse(A)
println("LU: ", xs)

# 方程式を解く
lu_solve(xs, b) |> println

LU: [3.0 1.0 1.0; 0.3333333333333333 1.6666666666666667 0.6666666666666667; 0.0 -0.6 1.4]
[-0.5714285714285714, -0.14285714285714302, 2.857142857142857]


## 乱数

Julia では `Random` 標準パッケージにて乱数関連のモジュールが提供されており、内部的には、高品質な乱数を高速に生成可能なメルセンヌ・ツイスタ アルゴリズムが使われている

In [24]:
using Random

# 0～1 の浮動小数点数を返す
rand()

0.3336167314392904

In [25]:
# 3x2 次元行列の形式で 0～1 の浮動小数点数を返す
rand(3, 2)

3×2 Matrix{Float64}:
 0.339792  0.62715
 0.99464   0.494605
 0.176197  0.902242

In [26]:
# 1～4 の範囲の整数を返す
rand(1:4)

3

In [27]:
# 2x3 次元行列の形式で 0～4 の範囲の整数を返す
rand(0:4, 2, 3)

2×3 Matrix{Int64}:
 1  1  2
 1  0  1

機械学習アルゴリズムでは、内部的に乱数を使うものがある

乱数を使うためある程度偶然により決まるのは仕方ないが、一方で同じデータに対しては同じ結果を返して欲しいということもよくある

そのような場合は乱数の種（シード）の設定を利用することになる

ここでわかり易い例として、サイコロを n 回振って出た目の和を返す関数を考える

In [28]:
throw_dice(n::Int) = rand(1:6, n) |> sum

# 10回サイコロを振って出た目の和を計算
throw_dice(10) |> println
throw_dice(10) |> println
throw_dice(10) |> println

35
44
39


上記のように、関数を呼び出すたびに結果が変わる

ここでは非常に単純な例を示しているが、もしこの関数 `throw_dice` が機械学習のアルゴリズムだとした場合、同じ引数（この場合 `n=10`）に対しては同じ結果を返してほしい場合がある

コンピュータの中で得られる乱数は擬似乱数と呼ばれるもので、これはある規則に基づく数列であるにも関わらず十分にランダムに見えるような工夫を施された数列である

その数列発生装置を初期化して最初から乱数を得ることができ、そうすることで何度でも同じ数列（乱数列）を繰り返すことが出来る

乱数列もいくつか選択肢があり、どの乱数列を選ぶかにあたるのが乱数の種（シード）と呼ばれるものである

乱数の種は通常整数値で与えられるが、同じ値で初期化すれば同じ乱数列が得られることが保証されている

In [29]:
# 乱数の種に 10 を指定
## ※ Juliaにおいては、グローバルな状態に影響を与えるような関数には慣例的に ! をつける
Random.seed!(10)

# 10回サイコロを振って出た目の和を計算
throw_dice(10) |> println

# サイコロを振る前に毎回乱数の種を 10 に初期化して throw_dice 関数を呼び出す
Random.seed!(10)
throw_dice(10) |> println

Random.seed!(10)
throw_dice(10) |> println

37
37
37


上記のように、毎回同じ結果が返ってくることがわかる

このように `Random.seed!` 関数を使うことで乱数列を再現することが出来るため、それを利用して結果が再現するように `throw_dice` 関数を書き換えてみる

In [30]:
throw_dice(n::Int; random_seed::Int = 10) = begin
    Random.seed!(random_seed)
    rand(1:6, n) |> sum
end

# 10回サイコロを振って出た目の和を計算
throw_dice(10) |> println
throw_dice(10) |> println
throw_dice(10) |> println

37
37
37


乱数を使うアルゴリズムが1つだけで、それを一度呼び出すだけであればこれで特に問題ないが、状況はもっと複雑なことが多い

例えば、機械学習のアルゴリズムが2種類あり、それらを少しずつ呼び出しながら最終的な結果を得るようなケースを考えてみる

こういった場合、乱数列はグローバルに1つだけ持つのでは都合が悪く、アルゴリズムの数だけ別々に乱数列を持つ方が都合が良い

それを実現するために Julia では `Random.MersenneTwister` 構造体が定義されており、`Random.rand` 関数の第1引数に渡すことで `Random.MersenneTwister` 構造体ごとに保有している異なる乱数列を参照・操作することが出来る

In [31]:
mutable struct Dice
    rnd::MersenneTwister # 乱数列
    sum::Int # 合計値
end

# Diceコンストラクタ: 以下の条件でインスタンス生成
## rnd = MersenneTwister(0): 乱数の種 0 で初期化された乱数列
## sum = 0
Dice() = Dice(MersenneTwister(0), 0)

# サイコロを振って、合計値に加算する
## Dice.rnd, Dice.sum 変数に影響を与えるため、関数名に ! をつけておくのが良い（慣例的に）
throw!(dice::Dice) = begin
    n = rand(dice.rnd, 1:6) # Dice構造体ごとに保有している rnd 変数（乱数列）を参照して 1～6 の乱数を生成
    dice.sum += n # 合計値に加算
end

throw! (generic function with 1 method)

In [32]:
"""
2つのサイコロを生成して同じ数だけ振れば、同じ結果になるはず
それを確かめる
"""

# サイコロ1
dice1 = Dice()

# サイコロ2
dice2 = Dice()

# サイコロ1、2ともに10回ずつ振る
for i in 1:10
    throw!(dice1)
    throw!(dice2)
end

# サイコロ1、2それぞれの合計値を確認
println(dice1.sum)
println(dice2.sum)

37
37
